From 24b6dc12a2c80e36b8ce6ebe4b5c44e51b582e43 Mon Sep 17 00:00:00 2001
From: linya14x <linx.yang@intel.com>
Date: Fri, 17 Apr 2026 14:58:31 +0800
Subject: [PATCH 07/28] media: add support lt6911uxc and enable split mode

2 pads linked to ipu 2 ports for split mode

single camera in bios setting, pprval for dummy port.
create software node and link connection in ipu-brige.
match the connection and port num in v4l2-async.
finally create the media links.

Signed-off-by: Chen Meng J <meng.j.chen@intel.com>
Signed-off-by: linya14x <linx.yang@intel.com>
---
 drivers/media/i2c/Kconfig                | 11 +++
 drivers/media/i2c/Makefile               |  1 +
 drivers/media/pci/intel/ipu-bridge.c     | 96 ++++++++++++++++++++----
 drivers/media/pci/intel/ipu6/ipu6-isys.c | 16 +++-
 drivers/media/v4l2-core/v4l2-async.c     | 16 +++-
 include/media/ipu-bridge.h               |  9 ++-
 include/media/v4l2-async.h               |  2 +
 7 files changed, 135 insertions(+), 16 deletions(-)

diff --git a/drivers/media/i2c/Kconfig b/drivers/media/i2c/Kconfig
index 3510b27f54..fe8b7345ce 100644
--- a/drivers/media/i2c/Kconfig
+++ b/drivers/media/i2c/Kconfig
@@ -1220,6 +1220,17 @@ config VIDEO_ISL7998X
 	  Support for Intersil ISL7998x analog to MIPI-CSI2 or
 	  BT.656 decoder.
 
+config VIDEO_LT6911UXC
+	tristate "Lontium LT6911UXC decoder"
+	depends on ACPI || COMPILE_TEST
+	select V4L2_CCI_I2C
+	help
+	  This is a Video4Linux2 sensor-level driver for the Lontium
+	  LT6911UXC HDMI to MIPI CSI-2 bridge.
+
+	  To compile this driver as a module, choose M here: the
+	  module will be called lt6911uxc.
+
 config VIDEO_LT6911UXE
 	tristate "Lontium LT6911UXE decoder"
 	depends on ACPI && VIDEO_DEV && I2C
diff --git a/drivers/media/i2c/Makefile b/drivers/media/i2c/Makefile
index 48b293a303..ace24bfabb 100644
--- a/drivers/media/i2c/Makefile
+++ b/drivers/media/i2c/Makefile
@@ -66,6 +66,7 @@ obj-$(CONFIG_VIDEO_ISL7998X) += isl7998x.o
 obj-$(CONFIG_VIDEO_KS0127) += ks0127.o
 obj-$(CONFIG_VIDEO_LM3560) += lm3560.o
 obj-$(CONFIG_VIDEO_LM3646) += lm3646.o
+obj-$(CONFIG_VIDEO_LT6911UXE) += lt6911uxc.o
 obj-$(CONFIG_VIDEO_LT6911UXE) += lt6911uxe.o
 obj-$(CONFIG_VIDEO_M52790) += m52790.o
 obj-$(CONFIG_VIDEO_MAX9271_LIB) += max9271.o
diff --git a/drivers/media/pci/intel/ipu-bridge.c b/drivers/media/pci/intel/ipu-bridge.c
index b1a7c7dcbf..d0c5df30aa 100644
--- a/drivers/media/pci/intel/ipu-bridge.c
+++ b/drivers/media/pci/intel/ipu-bridge.c
@@ -35,6 +35,11 @@
  */
 #define IVSC_DEV_NAME "intel_vsc"
 
+static const struct ipu_sensor_config ipu_supported_sensors_dummy[] = {
+	/* Lontium LT6911UXC dummy port for split mode */
+	IPU_SENSOR_CONFIG("INTC10B1", 0),
+};
+
 /*
  * Extend this array with ACPI Hardware IDs of devices known to be working
  * plus the number of link-frequencies expected by their drivers, along with
@@ -72,6 +77,8 @@ static const struct ipu_sensor_config ipu_supported_sensors[] = {
 	IPU_SENSOR_CONFIG("INT3537", 1, 437000000),
 	/* Lontium lt6911uxe */
 	IPU_SENSOR_CONFIG("INTC10C5", 0),
+	/* Lontium LT6911UXC */
+	IPU_SENSOR_CONFIG("INTC10B1", 0),
 	/* Omnivision OV01A10 / OV01A1S */
 	IPU_SENSOR_CONFIG("OVTI01A0", 1, 400000000),
 	IPU_SENSOR_CONFIG("OVTI01AS", 1, 400000000),
@@ -313,6 +320,7 @@ int ipu_bridge_parse_ssdb(struct acpi_device *adev, struct ipu_sensor *sensor)
 	}
 
 	sensor->link = ssdb.link;
+	sensor->pprval = ssdb.pprval;
 	sensor->lanes = ssdb.lanes;
 	sensor->mclkspeed = ssdb.mclkspeed;
 	sensor->rotation = ipu_bridge_parse_rotation(adev, &ssdb);
@@ -472,18 +480,49 @@ static void ipu_bridge_init_swnode_group(struct ipu_sensor *sensor)
 	}
 }
 
+static struct ipu_sensor *find_real_sensor(struct ipu_bridge *bridge,
+					   struct acpi_device *adev)
+{
+	struct ipu_sensor *sensor;
+	unsigned int i;
+
+	for (i = 0; i < bridge->n_sensors; i++) {
+		sensor = &bridge->sensors[i];
+		if (sensor->adev == ACPI_PTR(acpi_dev_get(adev)))
+			return sensor;
+	}
+
+	return NULL;
+}
+
 static void ipu_bridge_create_connection_swnodes(struct ipu_bridge *bridge,
-						 struct ipu_sensor *sensor)
+						 struct ipu_sensor *sensor,
+						 bool dummy)
 {
 	struct ipu_node_names *names = &sensor->node_names;
 	struct software_node *nodes = sensor->swnodes;
 
 	ipu_bridge_init_swnode_names(sensor);
 
-	nodes[SWNODE_SENSOR_HID] = NODE_SENSOR(sensor->name,
-					       sensor->dev_properties);
-	nodes[SWNODE_SENSOR_PORT] = NODE_PORT(sensor->node_names.port,
-					      &nodes[SWNODE_SENSOR_HID]);
+	if (dummy) {
+		struct ipu_sensor *real_sensor;
+
+		real_sensor = find_real_sensor(bridge, sensor->adev);
+		nodes[SWNODE_SENSOR_HID] =
+				real_sensor->swnodes[SWNODE_SENSOR_HID];
+		/* 1 for dummy port */
+		snprintf(sensor->node_names.port,
+				sizeof(sensor->node_names.port),
+				SWNODE_GRAPH_PORT_NAME_FMT, 1);
+		nodes[SWNODE_SENSOR_PORT] = NODE_PORT(sensor->node_names.port,
+				&real_sensor->swnodes[SWNODE_SENSOR_HID]);
+	} else {
+		nodes[SWNODE_SENSOR_HID] = NODE_SENSOR(sensor->name,
+				sensor->dev_properties);
+		nodes[SWNODE_SENSOR_PORT] = NODE_PORT(sensor->node_names.port,
+				&nodes[SWNODE_SENSOR_HID]);
+	}
+
 	nodes[SWNODE_SENSOR_ENDPOINT] = NODE_ENDPOINT(
 						sensor->node_names.endpoint,
 						&nodes[SWNODE_SENSOR_PORT],
@@ -662,7 +701,7 @@ static void ipu_bridge_unregister_sensors(struct ipu_bridge *bridge)
 }
 
 static int ipu_bridge_connect_sensor(const struct ipu_sensor_config *cfg,
-				     struct ipu_bridge *bridge)
+				     struct ipu_bridge *bridge, bool dummy)
 {
 	struct fwnode_handle *fwnode, *primary;
 	struct ipu_sensor *sensor;
@@ -685,31 +724,53 @@ static int ipu_bridge_connect_sensor(const struct ipu_sensor_config *cfg,
 		if (ret)
 			goto err_put_adev;
 
+		if (dummy) {
+			if (sensor->link == sensor->pprval) {
+				acpi_dev_put(adev);
+				continue;
+			} else {
+				dev_info(ADEV_DEV(adev),
+					 "dummy sensor, link %d, pprval %d\n",
+					 sensor->link, sensor->pprval);
+				sensor->link = sensor->pprval;
+			}
+		}
+
 		snprintf(sensor->name, sizeof(sensor->name), "%s-%u",
 			 cfg->hid, sensor->link);
-
 		ret = ipu_bridge_check_ivsc_dev(sensor, adev);
 		if (ret)
 			goto err_put_adev;
 
 		ipu_bridge_create_fwnode_properties(sensor, bridge, cfg);
-		ipu_bridge_create_connection_swnodes(bridge, sensor);
+		if (dummy) {
+			/* need adev to match the dummy port with the real one */
+			sensor->adev = adev;
+		}
+		ipu_bridge_create_connection_swnodes(bridge, sensor, dummy);
 
-		ret = software_node_register_node_group(sensor->group);
+		if (dummy) {
+			ret = software_node_register_node_group(&sensor->group[1]);
+		} else {
+			ret = software_node_register_node_group(sensor->group);
+		}
 		if (ret)
 			goto err_put_ivsc;
 
+		if (!dummy) {
 		fwnode = software_node_fwnode(&sensor->swnodes[
 						      SWNODE_SENSOR_HID]);
 		if (!fwnode) {
 			ret = -ENODEV;
 			goto err_free_swnodes;
 		}
+		}
 
 		sensor->adev = ACPI_PTR(acpi_dev_get(adev));
-
-		primary = acpi_fwnode_handle(adev);
-		primary->secondary = fwnode;
+		if (!dummy) {
+			primary = acpi_fwnode_handle(adev);
+			primary->secondary = fwnode;
+		}
 
 		ret = ipu_bridge_instantiate_ivsc(sensor);
 		if (ret)
@@ -742,7 +803,16 @@ static int ipu_bridge_connect_sensors(struct ipu_bridge *bridge)
 		const struct ipu_sensor_config *cfg =
 			&ipu_supported_sensors[i];
 
-		ret = ipu_bridge_connect_sensor(cfg, bridge);
+		ret = ipu_bridge_connect_sensor(cfg, bridge, 0);
+		if (ret)
+			goto err_unregister_sensors;
+	}
+
+	for (i = 0; i < ARRAY_SIZE(ipu_supported_sensors_dummy); i++) {
+		const struct ipu_sensor_config *cfg =
+			&ipu_supported_sensors_dummy[i];
+
+		ret = ipu_bridge_connect_sensor(cfg, bridge, 1);
 		if (ret)
 			goto err_unregister_sensors;
 	}
diff --git a/drivers/media/pci/intel/ipu6/ipu6-isys.c b/drivers/media/pci/intel/ipu6/ipu6-isys.c
index 48420e14df..72af4b485d 100644
--- a/drivers/media/pci/intel/ipu6/ipu6-isys.c
+++ b/drivers/media/pci/intel/ipu6/ipu6-isys.c
@@ -105,6 +105,7 @@ static int isys_isr_one(struct ipu6_bus_device *adev);
 static int
 isys_complete_ext_device_registration(struct ipu6_isys *isys,
 				      struct v4l2_subdev *sd,
+				      s16 src_pad,
 				      struct ipu6_isys_csi2_config *csi2)
 {
 	struct device *dev = &isys->adev->auxdev.dev;
@@ -122,6 +123,17 @@ isys_complete_ext_device_registration(struct ipu6_isys *isys,
 		goto unregister_subdev;
 	}
 
+	if (src_pad >= 0)
+		i = (unsigned int)src_pad;
+
+	if (sd->entity.pads[i].flags & MEDIA_PAD_FL_SOURCE) {
+		dev_info(dev, "src pad %d\n", src_pad);
+	} else {
+		dev_warn(dev, "src pad %d not for src\n", src_pad);
+		ret = -ENOENT;
+		goto unregister_subdev;
+	}
+
 	ret = media_create_pad_link(&sd->entity, i,
 				    &isys->csi2[csi2->port].asd.sd.entity,
 				    0, MEDIA_LNK_FL_ENABLED |
@@ -693,7 +705,9 @@ static int isys_notifier_bound(struct v4l2_async_notifier *notifier,
 
 	dev_dbg(&isys->adev->auxdev.dev, "bind %s nlanes is %d port is %d\n",
 		sd->name, s_asd->csi2.nlanes, s_asd->csi2.port);
-	ret = isys_complete_ext_device_registration(isys, sd, &s_asd->csi2);
+	ret = isys_complete_ext_device_registration(isys, sd,
+						    asc->match.src_pad,
+						    &s_asd->csi2);
 	if (ret)
 		return ret;
 
diff --git a/drivers/media/v4l2-core/v4l2-async.c b/drivers/media/v4l2-core/v4l2-async.c
index ee884a8221..57701be105 100644
--- a/drivers/media/v4l2-core/v4l2-async.c
+++ b/drivers/media/v4l2-core/v4l2-async.c
@@ -123,6 +123,18 @@ match_fwnode_one(struct v4l2_async_notifier *notifier,
 		"v4l2-async: device--endpoint match %sfound\n",
 		ret ? "" : "not ");
 
+	if (ret) {
+		struct fwnode_endpoint fwnode_ep = { 0 };
+		int ret_ep;
+		ret_ep = fwnode_graph_parse_endpoint(match->fwnode, &fwnode_ep);
+		if (ret_ep >= 0) {
+			match->src_pad = fwnode_ep.port;
+		} else {
+			match->src_pad = -1;
+		}
+
+	}
+
 	return ret;
 }
 
@@ -200,8 +212,10 @@ v4l2_async_find_match(struct v4l2_async_notifier *notifier,
 		}
 
 		/* match cannot be NULL here */
-		if (match(notifier, sd, &asc->match))
+		if (match(notifier, sd, &asc->match)) {
+			asc->src_pad = asc->match.src_pad;
 			return asc;
+		}
 	}
 
 	return NULL;
diff --git a/include/media/ipu-bridge.h b/include/media/ipu-bridge.h
index 16fac76545..42799f13d4 100644
--- a/include/media/ipu-bridge.h
+++ b/include/media/ipu-bridge.h
@@ -91,7 +91,13 @@ struct ipu_sensor_ssdb {
 	u8 controllogicid;
 	u8 reserved1[3];
 	u8 mclkport;
-	u8 reserved2[13];
+	u8 pmicpos;
+	u8 voltagerail;
+	u8 pprval;
+	u8 pprunit;
+	u8 flashid;
+	u8 phyconfig;
+	u8 reserved2[7];
 } __packed;
 
 struct ipu_property_names {
@@ -134,6 +140,7 @@ struct ipu_sensor {
 	struct ipu_node_names node_names;
 
 	u8 link;
+	u8 pprval;
 	u8 lanes;
 	u32 mclkspeed;
 	u32 rotation;
diff --git a/include/media/v4l2-async.h b/include/media/v4l2-async.h
index f26c323e9c..3d1573a7bf 100644
--- a/include/media/v4l2-async.h
+++ b/include/media/v4l2-async.h
@@ -59,6 +59,7 @@ struct v4l2_async_match_desc {
 			unsigned short address;
 		} i2c;
 	};
+	s16 src_pad;
 };
 
 /**
@@ -82,6 +83,7 @@ struct v4l2_async_connection {
 	struct list_head asc_entry;
 	struct list_head asc_subdev_entry;
 	struct v4l2_subdev *sd;
+	s16 src_pad;
 };
 
 /**
-- 
2.53.0

